### PR TITLE
Resolve nested seed paths in v01 PDA extraction

### PR DESCRIPTION
**Pull request by ioxde**
This is a direct follow-up to #984 which left the nested path handling as a TODO.

PDA seeds with nested paths like `mint.authority` or `args.owner` were skipped. This PR resolves them by looking up the field type from IDL type definitions.

`{ kind: 'account', path: 'mint.authority' }` + a `Mint` type def with `authority: pubkey` -> `variablePdaSeedNode('mintAuthority', publicKeyTypeNode())`

`{ kind: 'arg', path: 'args.owner' }` where `args` is a struct with `owner: pubkey` -> `variablePdaSeedNode('owner', publicKeyTypeNode())`

Program seeds with nested paths still skip (with a warning). `pdaSeedNodeFromAnchorV01` can now return `undefined`.

### CODE REVIEW

**Comment by lorisleiva** [comment]
@trevor-cortex

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts [hunk 1]
```diff
@@ -34,40 +40,115 @@ export function pdaSeedNodeFromAnchorV01(
...
+            if (
+                isNode(argumentType, 'sizePrefixTypeNode') &&
+                isNode(argumentType.type, 'stringTypeNode') &&
+                argumentType.type.encoding === 'utf8' &&
+                isNode(argumentType.prefix, 'numberTypeNode') &&
+                argumentType.prefix.format === 'u32'
+            ) {
+                argumentType = stringTypeNode('utf8');
+            }
+
```
Comment (trevor-cortex) [inline packages/nodes-from-anchor/src/v01/PdaSeedNode.ts:102]: Nit: the two return branches at line 102–106 and 109–112 produce the same shape. You could collapse them into a single return after the string-unwrapping logic:

```ts
return {
    definition: variablePdaSeedNode(argumentName, argumentType),
    value: pdaSeedValueNode(argumentName, argumentValueNode(argumentName)),
};
```

The `if (pathParts.length > 1)` guard before this return isn't doing anything differently.

File: packages/nodes-from-anchor/src/v01/InstructionAccountNode.ts [hunk 2]
```diff
@@ -74,38 +88,61 @@ export function instructionAccountNodeFromAnchorV01(
...
-                seedValues,
-                programIdValue,
-            );
+                const camelName = camelCase(name);
+                const isSelfReferential =
+                    seedValues.some(sv => isNode(sv.value, 'accountValueNode') && sv.value.name === camelName) ||
+                    (programIdValue != null &&
+                        isNode(programIdValue, 'accountValueNode') &&
+                        programIdValue.name === camelName);
+                if (isSelfReferential) {
```
Comment (trevor-cortex) [inline packages/nodes-from-anchor/src/v01/InstructionAccountNode.ts]: Ultra-minor: the `if (isSelfReferential)` warning and `if (!isSelfReferential)` assignment could be a single `if/else`:

```ts
if (isSelfReferential) {
    logWarn(...);
} else {
    defaultValue = pdaValueNode(...);
}
```

**Review by trevor-cortex — APPROVED** [review]
## Summary

This PR resolves nested seed paths (e.g. `mint.authority`, `args.owner`) in v01 PDA extraction, which were previously left as a TODO and silently skipped. It introduces a `resolveNestedFieldType` helper that walks struct fields (inline or via IDL type definitions) to determine the leaf type. The function `pdaSeedNodeFromAnchorV01` can now return `undefined` when resolution fails, and callers handle this gracefully by skipping the PDA default value.

Additionally, the PR adds self-referential seed detection — if a seed references the account being derived, the PDA default is skipped with a warning.

## Key things to watch

1. **Breaking return type change**: `pdaSeedNodeFromAnchorV01` now returns `| undefined`. This is a well-handled breaking change internally — all call sites check for `undefined` — but any external consumers of this export would need updating.

2. **Self-referential detection**: The new self-referential check (lines 131–138 in `InstructionAccountNode.ts`) correctly catches cases like `vault` referencing itself, and nested paths like `guard.mint` where the account name matches the seed's account reference. The test for `my_guard` with `path: 'my_guard.mint'` is a good edge case.

3. **Nested program paths**: Program seeds with nested paths are still explicitly skipped with a warning — this is a reasonable incremental choice since program seeds with nested paths are rare.

## Notes for subsequent reviewers

- The `resolveNestedFieldType` helper only handles `structTypeNode` and `definedTypeLinkNode`. Other wrappers (e.g. `optionTypeNode`) would fall through and return `undefined`. This seems fine for now — real-world IDL seeds shouldn't reference optional fields — but worth keeping in mind.
- The `InstructionNode.test.ts` change for the `distribution` account is interesting: it confirms the self-referential detection works at the integration level (the account `distribution` has a seed `path: 'distribution.group_mint'` referencing itself, so the PDA default is correctly omitted).
- Test coverage is thorough: nested account paths, nested arg paths, unresolvable types, self-referential accounts, prefixed nested groups, and program seed edge cases are all covered.

File: packages/nodes-from-anchor/test/v01/InstructionAccountNode.test.ts [hunk 3]
```diff
@@ -291,17 +294,75 @@ test('it correctly prefixes PDA seed account references in nested groups', () =>
...
     const nodes = instructionAccountNodesFromAnchorV01(
-        // [
-        //     accountNode({
-        //         data: sizePrefixTypeNode(
-        //             structTypeNode([structFieldTypeNode({ name: 'authority', type: publicKeyTypeNode() })]),
-        //             numberTypeNode('u32'),
-        //         ),
-        //         name: 'mint',
-        //     }),
-        // ],
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/test/v01/InstructionAccountNode.test.ts:304]: The idea was that these would be uncommented when nested arguments/accounts are supported. Is the idea that the new tests are taking care of that deleted code now?

File: packages/nodes-from-anchor/test/v01/InstructionAccountNode.test.ts [hunk 4]
```diff
@@ -406,17 +596,99 @@ test('it handles PDAs with a program id that points to another account', () => {
...
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/test/v01/InstructionAccountNode.test.ts:419]: Same here

File: packages/nodes-from-anchor/src/v01/ProgramNode.ts [hunk 5]
```diff
@@ -27,7 +27,7 @@ export function programNodeFromAnchorV01(idl: IdlV01): ProgramNode {
         definedTypes,
         errors: errors.map(errorNodeFromAnchorV01),
         events: events.map(event => eventNodeFromAnchorV01(event, types, generics)),
-        instructions: instructions.map(instruction => instructionNodeFromAnchorV01(instruction, generics)),
+        instructions: instructions.map(instruction => instructionNodeFromAnchorV01(instruction, generics, types)),
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/src/v01/ProgramNode.ts]: nit: `instructionNodeFromAnchorV01(instruction, types, generics)`. All other functions order them like that.

File: packages/nodes-from-anchor/src/v01/InstructionAccountNode.ts [hunk 6]
```diff
@@ -74,38 +88,61 @@ export function instructionAccountNodeFromAnchorV01(
     if (idl.address) {
         defaultValue = publicKeyValueNode(idl.address, name);
     } else if (idl.pda) {
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/src/v01/InstructionAccountNode.ts:91]: This is starting to have way too much code for an if-statement. Would you mind extracting this to a local private function so we can understand this better?

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts [hunk 7]
```diff
@@ -34,40 +40,115 @@ export function pdaSeedNodeFromAnchorV01(
...
             const prefixedAccountName = prefix ? `${prefix}_${accountName}` : accountName;
+
+            if (pathParts.length > 1) {
+                const accountTypeName = seed.account ?? accountName;
+                const rootType = typeNodeFromAnchorV01({ defined: { name: accountTypeName } }, generics);
+                const resolved = resolveNestedFieldType(rootType, pathParts.slice(1), idlTypes, generics);
+                if (!resolved) {
+                    logWarn(`Could not resolve nested account path "${seed.path}" for PDA seed.`);
+                    return undefined;
+                }
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/src/v01/PdaSeedNode.ts:54]: I wonder if here we should just throw? Before we were silently ignoring nested paths because that was a valid scenario we were not able to support. Now that we are, if the nested seed don't resolve properly than it is a malformation on the IDL itself right?

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts [hunk 8]
```diff
@@ -34,40 +40,115 @@ export function pdaSeedNodeFromAnchorV01(
...
         }
         case 'arg': {
-            // Ignore nested paths.
-            const [originalArgumentName] = seed.path.split('.');
-            const argumentName = camelCase(originalArgumentName);
-            const argumentNode = instructionArguments.find(({ name }) => name === argumentName);
-            if (!argumentNode) {
-                throw new CodamaError(CODAMA_ERROR__ANCHOR__ARGUMENT_TYPE_MISSING, { name: originalArgumentName });
+            const pathParts = seed.path.split('.');
+            const argumentName = camelCase(pathParts.length > 1 ? pathParts[pathParts.length - 1] : pathParts[0]);
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/src/v01/PdaSeedNode.ts]: nit: this is equivalent to:


```suggestion
            const argumentName = camelCase(pathParts[pathParts.length - 1]);
```

Because `pathParts.length - 1 === 0` when `pathParts.length === 1`.

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts [hunk 9]
```diff
@@ -34,40 +40,115 @@ export function pdaSeedNodeFromAnchorV01(
...
 
-            // Anchor uses unprefixed strings for PDA seeds even though the
-            // argument itself uses a Borsh size-prefixed string. Thus, we
-            // must recognize this case and convert the type accordingly.
-            const isBorshString =
-                isNode(argumentNode.type, 'sizePrefixTypeNode') &&
-                isNode(argumentNode.type.type, 'stringTypeNode') &&
-                argumentNode.type.type.encoding === 'utf8' &&
-                isNode(argumentNode.type.prefix, 'numberTypeNode') &&
-                argumentNode.type.prefix.format === 'u32';
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/src/v01/PdaSeedNode.ts:62]: I'd rather you kept the `isBorshString` variable so we know what this massive if condition checks.

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts [hunk 10]
```diff
@@ -34,40 +40,115 @@ export function pdaSeedNodeFromAnchorV01(
...
             };
         }
         default:
             throw new CodamaError(CODAMA_ERROR__ANCHOR__SEED_KIND_UNIMPLEMENTED, { kind });
     }
 }
+
+function resolveNestedFieldType(
+    rootType: TypeNode,
+    fieldPath: string[],
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/src/v01/PdaSeedNode.ts]: nit: Keep your naming consistent.

```suggestion
    pathParts: string[],
```

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts [hunk 11]
```diff
@@ -34,40 +40,115 @@ export function pdaSeedNodeFromAnchorV01(
...
+    idlTypes: IdlV01TypeDef[],
+    generics: GenericsV01,
+): TypeNode | undefined {
+    let currentType = rootType;
+
+    for (const fieldName of fieldPath) {
+        const target = camelCase(fieldName);
+
+        if (isNode(currentType, 'structTypeNode')) {
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/src/v01/PdaSeedNode.ts:132]: Do we also need to worry about `tupleTypeNodes`? Is having a path like `foo.0.bar` a thing?

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts [hunk 12]
```diff
@@ -34,40 +40,115 @@ export function pdaSeedNodeFromAnchorV01(
...
+                value: pdaSeedValueNode(argumentName, argumentValueNode(argumentName)),
             };
         }
         default:
             throw new CodamaError(CODAMA_ERROR__ANCHOR__SEED_KIND_UNIMPLEMENTED, { kind });
     }
 }
+
+function resolveNestedFieldType(
+    rootType: TypeNode,
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/src/v01/PdaSeedNode.ts]: I wouldn't call that `rootType` since there is a `RootNode` which makes this confusing.

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts [hunk 13]
```diff
@@ -34,40 +40,115 @@ export function pdaSeedNodeFromAnchorV01(
...
+            const linkName = currentType.name;
+            const typeDef = idlTypes.find(t => camelCase(t.name) === linkName);
+            if (!typeDef || typeDef.type.kind !== 'struct' || !typeDef.type.fields) return undefined;
+            const { fields } = typeDef.type;
+            if (!fields.length || typeof fields[0] !== 'object' || !('name' in fields[0])) return undefined;
+            const field = (fields as IdlV01Field[]).find(f => camelCase(f.name) === target);
+            if (!field) return undefined;
+            currentType = typeNodeFromAnchorV01(field.type, generics);
+            continue;
+        }
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/src/v01/PdaSeedNode.ts]: Could you not prefix this before the `structTypeNode` condition and do something like this instead;

```ts
if (isNode(currentType, 'definedTypeLinkNode')) {
    const linkName = currentType.name;
    const typeDef = idlTypes.find(t => camelCase(t.name) === linkName);
    if (!typeDef) return undefined;
    currentType = typeNodeFromAnchorV01(typeDef, generics);
}

// Now the `structTypeNode` condition (and any others in the future) can pick up from that fetched linked node.
```

File: packages/nodes-from-anchor/test/v01/InstructionNode.test.ts [hunk 14]
```diff
@@ -34,23 +34,13 @@ test('it creates instruction nodes', () => {
...
-                    // defaultValue: pdaValueNode(
-                    //     pdaNode({
-                    //         name: 'distribution',
-                    //         seeds: [
-                    //             constantPdaSeedNodeFromBytes('base58', 'F9bS'),
-                    //             variablePdaSeedNode('distributionGroupMint', publicKeyTypeNode()),
-                    //         ],
-                    //     }),
-                    //     [pdaSeedValueNode("distributionGroupMint", accountValueNode('distribution', 'group_mint'))],
-                    // ),
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/test/v01/InstructionNode.test.ts:53]: Same here

File: packages/nodes-from-anchor/test/v01/pdaSeedNode.test.ts [hunk 15]
```diff
@@ -17,24 +20,86 @@ import { pdaSeedNodeFromAnchorV01 } from '../../src';
 test('it creates a PdaSeedNode from a const Anchor seed', () => {
     const nodes = pdaSeedNodeFromAnchorV01({ kind: 'const', value: [11, 57, 246, 240] }, []);
 
-    expect(nodes.definition).toEqual(constantPdaSeedNodeFromBytes('base58', 'HeLLo'));
-    expect(nodes.value).toBeUndefined();
+    expect(nodes!.definition).toEqual(constantPdaSeedNodeFromBytes('base58', 'HeLLo'));
+    expect(nodes!.value).toBeUndefined();
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/test/v01/pdaSeedNode.test.ts]: If we throw instead of returning undefined, we remove the need for these casts.

**Review by lorisleiva — CHANGES_REQUESTED** [review]

File: packages/nodes-from-anchor/test/v01/InstructionAccountNode.test.ts (hunk 3 above)
Comment (ioxde) [reply packages/nodes-from-anchor/test/v01/InstructionAccountNode.test.ts:304]: The commented code assumed account data would be passed directly via `accountNode`, but nested types are now resolved through `idlTypes` instead.

File: packages/nodes-from-anchor/test/v01/InstructionAccountNode.test.ts (hunk 4 above)
Comment (ioxde) [reply packages/nodes-from-anchor/test/v01/InstructionAccountNode.test.ts:419]: Same as above

File: packages/nodes-from-anchor/src/v01/InstructionAccountNode.ts (hunk 6 above)
Comment (ioxde) [reply packages/nodes-from-anchor/src/v01/InstructionAccountNode.ts:91]: Good idea. This comment let me also catch a mistake [if a program seed has a nested path](https://github.com/ioxde/codama/commit/10ab1a84e7e6e39e9fd5b3280b7f44e43076ddb1) being skipped.

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts (hunk 8 above)
Comment (ioxde) [reply packages/nodes-from-anchor/src/v01/PdaSeedNode.ts]: This was changed in a37264c. It was changed to `pathParts.slice(1).join('_')` to avoid collisions so the ternary is needed now. The rebase onto 394e9b6a exposed the issue.

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts (hunk 12 above)
Comment (ioxde) [reply packages/nodes-from-anchor/src/v01/PdaSeedNode.ts]: `parentType` okay?

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts (hunk 13 above)
Comment (ioxde) [reply packages/nodes-from-anchor/src/v01/PdaSeedNode.ts]: Ended up changing this with feedback from above

File: packages/nodes-from-anchor/test/v01/InstructionNode.test.ts (hunk 14 above)
Comment (ioxde) [reply packages/nodes-from-anchor/test/v01/InstructionNode.test.ts:53]: Same feedback about using `idlTypes` as above.

File: packages/nodes-from-anchor/test/v01/pdaSeedNode.test.ts (hunk 15 above)
Comment (ioxde) [reply packages/nodes-from-anchor/test/v01/pdaSeedNode.test.ts]: I switched all tests to `?.`.  Unsure about the `throw`. See above comment about throw vs log.

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts (hunk 7 above)
Comment (ioxde) [reply packages/nodes-from-anchor/src/v01/PdaSeedNode.ts:54]: Honestly not sure. I know Anchor PDAs aren't perfect. I think the best path would be to throw by default (with a detailed message pointing to docs/what to fix) but allowing a user to override per error. I am not sure that whole change fits in this PR though. Seems more like a project-level standard / convention?

I personally do like the fail-first override behavior because it forces the user to understand their decisions. Console warnings don't always tell the real severity of the issue / easy to accidentally ignore.

I'll wait for more direction to make changes here.

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts (hunk 11 above)
Comment (ioxde) [reply packages/nodes-from-anchor/src/v01/PdaSeedNode.ts:132]: Great catch b0c4d41b1b39a85eca83c93349afa1b9035b3568

**Comment by ioxde** [comment]
I think I addressed all comments outside of the `logWarn` vs `throw` question. Let me know if you have any other questions / concerns.

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts (hunk 7 above)
Comment (lorisleiva) [reply packages/nodes-from-anchor/src/v01/PdaSeedNode.ts:54]: I'm happy with keeping this as-is if you're concerned about malformed Anchor IDLs.

File: packages/nodes-from-anchor/test/v01/pdaSeedNode.test.ts [hunk 16]
```diff
@@ -17,24 +21,151 @@ import { pdaSeedNodeFromAnchorV01 } from '../../src';
...
+test('it resolves nested path through tuple then struct (foo.0.bar)', () => {
+    const nodes = pdaSeedNodeFromAnchorV01({ kind: 'arg', path: 'foo.0.bar' }, [
+        instructionArgumentNode({
+            name: 'foo',
+            type: tupleTypeNode([structTypeNode([structFieldTypeNode({ name: 'bar', type: numberTypeNode('u8') })])]),
+        }),
+    ]);
+
+    expect(nodes?.definition).toEqual(variablePdaSeedNode('0Bar', numberTypeNode('u8')));
+    expect(nodes?.value).toEqual(pdaSeedValueNode('0Bar', argumentValueNode('0Bar')));
```
Comment (lorisleiva) [inline packages/nodes-from-anchor/test/v01/pdaSeedNode.test.ts:148]: This worries me a bit because most languages will refuse to render a variable that starts with a numerical value. I wonder if we should just use the whole path as the seed name. So here it would become `foo0Bar`. That way, it would technically not be possible for a variable to start with a number because we always start with the array of `InstructionArgumentNodes`.

**Review by lorisleiva — CHANGES_REQUESTED** [review]
Thank you for all the changes and sorry for the late re-review.

Just one last little thing that caught my attention and after that this is good to go. Thank you for your patience. 🙏

File: packages/dynamic-client/test/programs/anchor/tests/blog.test.ts [hunk 17]
```diff
@@ -10,7 +10,7 @@ import { loadIdl, SvmTestContext } from '../../test-utils';
 
 const idl = loadIdl('blog-idl.json');
 const programClient = createProgramClient<BlogProgramClient>(idl);
-const programSoPath = path.resolve(__dirname, '..', '..', 'dumps', 'blog.so');
+const programSoPath = path.resolve(__dirname, '..', 'target', 'deploy', 'blog.so');
```
Comment (mikhd) [inline packages/dynamic-client/test/programs/anchor/tests/blog.test.ts:13]: Hi @ioxde,

I think it should be resolved `.so` from dumps.

As recently we've [get rid](https://github.com/codama-idl/codama/pull/1005) of custom anchor programs building in tests and instead tests use pre-build committed binaries from dumps.

Thank you!

File: packages/nodes-from-anchor/src/v01/PdaSeedNode.ts [hunk 18]
```diff
@@ -34,40 +40,111 @@ export function pdaSeedNodeFromAnchorV01(
...
         }
         case 'arg': {
-            // Ignore nested paths.
-            const [originalArgumentName] = seed.path.split('.');
-            const argumentName = camelCase(originalArgumentName);
-            const argumentNode = instructionArguments.find(({ name }) => name === argumentName);
-            if (!argumentNode) {
-                throw new CodamaError(CODAMA_ERROR__ANCHOR__ARGUMENT_TYPE_MISSING, { name: originalArgumentName });
+            const pathParts = seed.path.split('.');
+            const argumentName = camelCase(pathParts.length > 1 ? pathParts.slice(1).join('_') : pathParts[0]);
```
Comment (mikhd) [inline packages/nodes-from-anchor/src/v01/PdaSeedNode.ts:69]: Could we double-check this, please?

The arg branch drops the root, so two different struct arguments sharing a field name within the same instruction always collide.

Lets say two different InstructionArgument inputs have the same keys.
If I understood this correctly, PdaSeed depending either on `input1.owner` or `input2.owner` will get the same `owner` name. So it will become impossible to identify pda seed dependency.

```
[
    instructionArgumentNode({
        name: 'input1',
        type: structTypeNode([
            structFieldTypeNode({ name: 'owner', type: publicKeyTypeNode() }),
            structFieldTypeNode({ name: 'amount', type: numberTypeNode('u64') }),
        ]),
    }),
    instructionArgumentNode({
        name: 'input2',
        type: structTypeNode([
            structFieldTypeNode({ name: 'owner', type: publicKeyTypeNode() }),
            structFieldTypeNode({ name: 'amount', type: numberTypeNode('u64') }),
        ]),
    }),
]
```
